include/boost/json/value_builder.hpp

```diff
@@ -0,0 +1,556 @@
+//
+// Copyright (c) 2019 Vinnie Falco (vinnie.falco@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/cppalliance/json
+//
+
+#ifndef BOOST_JSON_VALUE_BUILDER_HPP
+#define BOOST_JSON_VALUE_BUILDER_HPP
+
+#include <boost/json/detail/config.hpp>
+#include <boost/json/error.hpp>
+#include <boost/json/storage_ptr.hpp>
+#include <boost/json/value.hpp>
+#include <boost/json/detail/raw_stack.hpp>
+#include <stddef.h>
+
+namespace boost {
+namespace json {
+
+//----------------------------------------------------------
+
+/** A factory for building a value DOM.
+
+    A value builder implements an algorithm for
+    efficiently constructing a @ref value from an
+    external source (provided by the caller).
+
+    The builder uses a dynamically allocated internal
+    storage to hold portions of the document, allowing
+    complete objects and arrays to be constructed using
+    a single allocation when their contents are
+    eventually known. This internal storage is reused
+    when creating multiple values with the same builder.
+
+    To use the builder construct it with an optionally
+    specified memory resource to use for the internal
+    storage. Then call @ref reset once before building
+    each complete DOM, optionally specifying the
+    memory resource to use for the resulting @ref value.
+
+    The functions @ref on_document_begin and
+    @ref on_document_end must be called exactly once
+    at the beginning and at the end of construction.
+    The remaining event handling functions are called
+    according to their descriptions to build the document.
+*/
+class value_builder
+{
+    enum class state : char;
+    struct level
+    {
+        std::uint32_t count;
+        char align;
+        state st;
+    };
+
+    storage_ptr sp_;
+    detail::raw_stack rs_;
+    std::uint32_t key_size_ = 0;
+    std::uint32_t str_size_ = 0;
+    level lev_;
+
+public:
+    /** Destructor.
+
+        All dynamically allocated memory, including
+        any partially built results, is freed.
+    */
+    BOOST_JSON_DECL
+    virtual
```

Comment:
We moved away from the inheritance model, should this still be virtual?

Reply:
Nope